runtime/hydradx/src/assets.rs

```diff
@@ -79,12 +85,128 @@ pub struct CurrencyHooks;
 impl MutationHooks<AccountId, AssetId, Balance> for CurrencyHooks {
 	type OnDust = Duster;
 	type OnSlash = ();
-	type PreDeposit = ();
+	type PreDeposit = SufficiencyCheck;
 	type PostDeposit = ();
-	type PreTransfer = ();
+	type PreTransfer = SufficiencyCheck;
 	type PostTransfer = ();
 	type OnNewTokenAccount = AddTxAssetOnAccount<Runtime>;
-	type OnKilledTokenAccount = RemoveTxAssetOnKilled<Runtime>;
+	type OnKilledTokenAccount = (RemoveTxAssetOnKilled<Runtime>, OnKilledTokenAccount);
+}
+
+pub const SUFFICIENCY_LOCK: LockIdentifier = *b"insuffEd";
+
+parameter_types! {
+	pub InsufficientEDinHDX: Balance = FixedU128::from_rational(11, 10)
+		.saturating_mul_int(<Runtime as pallet_balances::Config>::ExistentialDeposit::get());
+}
+
+pub struct SufficiencyCheck;
+impl SufficiencyCheck {
+	fn on_funds(asset: AssetId, paying_account: &AccountId) -> DispatchResult {
+		if <Runtime as orml_tokens::Config>::DustRemovalWhitelist::contains(paying_account) {
+			return Ok(());
+		}
+
+		//NOTE: account existence means it already paid ED
+		if orml_tokens::Accounts::<Runtime>::try_get(paying_account, asset).is_err()
+			&& !AssetRegistry::is_sufficient(asset)
+		{
+			let fee_payment_asset = MultiTransactionPayment::account_currency(paying_account);
+
+			//NOTE: unwrap should never fail. MutiTransctionPayment should always provide price.
```

Comment:
This is not true. We won't have price for newly created assets

Reply:
These assets can't be used as fee asset, you can use only accepted currencies https://github.com/galacticcouncil/HydraDX-node/blob/2cc7a5fec807744d382689d5847172bac50008c5/pallets/transaction-multi-payment/src/lib.rs#L239-L242 for which you have to provide a price when you are adding it into accepted currencies.

Based on this comment we always provides some kind of fallback price:
https://github.com/galacticcouncil/HydraDX-node/blob/2cc7a5fec807744d382689d5847172bac50008c5/pallets/transaction-multi-payment/src/lib.rs#L186

But you are right, I didn't think about case when currency was removed from accepted currencies and user still have it set as fee asset, I'll make it fail with:
https://github.com/galacticcouncil/HydraDX-node/blob/2cc7a5fec807744d382689d5847172bac50008c5/pallets/transaction-multi-payment/src/lib.rs#L155